utils/initial_data_preparation.py

```diff
@@ -68,88 +80,104 @@ def parse_args():
     """
-    logger.info("Starting to remove duplicate rows.")
-    initial_count = len(df)
-    duplicates = df.duplicated().sum()
-
-    if duplicates:
-        logger.info(f"Number of duplicate rows: {duplicates}")
-        # Keep the first occurrence of each duplicate row and remove the others
-        deduped_df = df.drop_duplicates(keep="first")
-        duplicates_removed = initial_count - len(deduped_df)
-        logger.info(
-            f"First occurrence of each duplicate row has been kept. "
-            f"Dropped {duplicates_removed} duplicate rows."
-        )
-    else:
-        logger.info("No duplicate rows found.")
-        deduped_df = df.copy()
-        duplicates_removed = 0  # No duplicates found, so no rows removed
 
-    return deduped_df, duplicates_removed
+    logger.info("Marking duplicate rows.")
 
+    # 'duplicate_flag' is `False` for the first occurrence, and it is `True`
+    # for the duplicate row
+    df["duplicate_flag"] = df.duplicated(keep="first")
+    duplicates_count = df["duplicate_flag"].sum()
 
-def remove_null_values(df):
+    logger.info(f"Marked {duplicates_count} duplicate rows.")
+
+    return df
+
+
+def mark_null_values(d):
     """
-    Removes rows with null values from the DataFrame and logs the result.
+    Adds a 'null_value_flag' column to the DataFrame to indicate rows with null
+    values.
 
     Parameters:
-        df (pd.DataFrame): The DataFrame from which to remove nulls.
+        df (pd.DataFrame): The DataFrame in which to mark null values.
 
     Returns:
-        tuple: A tuple containing the non-null DataFrame and the count of
-        removed rows.
+        pd.DataFrame: The DataFrame with an added 'null_value_flag' column.
     """
-    initial_count = len(df)
-    null_counts = df.isnull().sum()
+    logger.info("Marking rows with null values.")
+
+    # Create a flag column for rows containing null values
+    df["null_value_flag"] = df.isnull().any(axis=1)
+    nulls_count = df["null_value_flag"].sum()
+
+    logger.info(f"Marked {nulls_count} rows with null values.")
+
+    return df
+
+
+# Define a function to extract domain and flag errors
+def extract_and_flag_domains(df):
+    """
+    Extracts domains from URLs and flags rows with unsupported URL schemes.
+
+    Parameters:
+        df (pd.DataFrame): DataFrame containing the 'repourl' column with URLs.
+
+    Returns:
+        pd.DataFrame: Updated DataFrame with 'repodomain' and 'unsupported_url_flag' columns.
+    """
+
+    def get_domain(url):
+        parsed_url = urlparse(url)
+        if parsed_url.scheme in ["http", "https", "git"]:
+            return parsed_url.hostname
+        else:
+            return None
+
+    # Apply the domain extraction function
+    df["repodomain"] = df["repourl"].apply(get_domain)
+
+    # Flag rows where the domain could not be extracted
+    # (i.e., unsupported schemes)
+    df["domain_extraction_flag"] = df["repodomain"].isnull()
```

Comment:
Here's an example where the flag's name might communicate the intention of the variable (the column) more clearly; and/or there may be a mistake in this line of code.

The scheme couldn't be extracted by our code (line 142 in this file) rather than the domain being extracted. From line 137 (part of the docstring) it says `unsupported_url_flag` columns is returned not `domain_extraction_flag`. Should this code be setting `unsupported_url_flag` instead?

Reply:
PS: I prefer `unsupported_url_scheme` as the column name as I think that is clearer than either of these two columns names.

Reply:
As the name of the function is `extract_and_flag_domains` and in the helper function `get_domain` we are trying to extract the `hostname`, I thought adding `domain_extraction_flag` would be a better choice. However, I agree that `unsupported_url_scheme` is a more precise choice. And I will change the docstring as well.
Thanks for your feedback.

Reply:
- Renamed column 'domain_extraction_flag' to 'unsupported_url_scheme' throughout the codebase as per review feedback.
- Updated references and function calls accordingly to reflect the new column name.